### PR TITLE
Support ufunc.outer

### LINKED ISSUE #2342: Implement outer ufuncs
The ufuncs in numpy all have useful `outer` methods which are useful and easily parallelizable.
For instance, `np.subtract.outer(x,y)` can be done in chunks of rows of length `y.size`.
`delayed(np.subtract.outer)` does not work because numpy casts the inputs to ndarrays.

Is this enhancement possible?

### PR DESCRIPTION
**Pull request by jcrist**
- Add support for `ufunc.outer` methods for dask versions
- Separate ufunc objects from wrapped elemwise functions. Ufuncs are now
  an instance of `da.ufunc.ufunc`, and have a number of attribute and
  methods mirrored from their numpy counterparts
- Fix a bug in `da.ufunc.angle`
- Improve test coverage for `da.ufunc`

Fixes #2342.

### CODE REVIEW

**Comment by jcrist** [comment]
Failures are due to ` ichart.yahoo.com` being down, resulting in an unrelated test failure.

**Comment by jcrist** [comment]
Planning to merge later today if no comment.

File: dask/array/tests/test_ufunc.py [hunk 1]
```diff
@@ -94,6 +109,44 @@ def test_ufunc_2args(ufunc):
...
+def test_ufunc_outer():
+    arr1 = np.random.randint(1, 100, size=20)
+    darr1 = da.from_array(arr1, 3)
+
+    arr2 = np.random.randint(1, 100, size=(10, 3))
+    darr2 = da.from_array(arr2, 3)
+
+    # Check output types
+    assert isinstance(da.add.outer(darr1, darr2), da.Array)
+    assert isinstance(da.add.outer(arr1, darr2), da.Array)
```
Comment (mrocklin) [inline dask/array/tests/test_ufunc.py]: `darr1, arr2`?

Comment (jcrist) [inline dask/array/tests/test_ufunc.py]: Addressed

**Comment by jcrist** [comment]
> I'm curious, what else might we want to add to ufunc in the future other than `.outer`?

The only other ones are `reduce`, `accumulate`, `reduceat`, and `at`. Since `at` is inplace, we can nix that one. The `reduce` ones may be useful, but since we're working with general ufuncs we'd have to perform a naive linear reduce (rather than using smarter, more efficient methods). Of course we could special case for known functions (e.g. `add`). `accumulate` would be more straightforward, and would look like the existing cumulative functions.

**Pull request merged**